text/temp/PALENICA/Larsson_Sadakane_Suffix_Array.tex

```diff
@@ -0,0 +1,105 @@
+
+Algorytm tworzenia tablicy sufiksowej zaprezentowany przez N. Larssona i K. Sadakane działa w czasie $O(n\ log\ n)$ gdzie $n$ jest długością wejśćiowego tekstu. Autorzy rozbudowują podejście \cite{basicAlgo} i korzystają z techniki `podwajania` znanej z algorytmu Karpa, Millera, Rosenbera która polega na użyciu pozycji sufiksów po fazie sortowania jako kluczy do sortowania dwukrotnie dłuższych sufiksów w kolejnej fazie. Autorzy formalizują to podejście definiując pojęcie \textit{h-order}, czyli porządek leksykograficzny sufiksów rozważając jedynie $h$ początkowych znaków każdego sufiksu. Zayważmy, że \textit{h-order} niekoniecznie jest jednoznaczny jeśli $h < n$.
+\begin{observation}
+Sortowanie sufiksów przy użyciu klucza w postaci pary złożonej z pozycji sufiksu $S_i$ w \textit{h-order} oraz pozycji sufiksu $S_{i+h}$ w \textit{h-order} daje sortowanie sufiksów w \textit{2h-order}.
+\end{observation}
+
+\begin{definition}
+Dla tablicy sufiksów $I$ która jest w \textit{h-order} definiujemy
+\begin{itemize}
+    \item Maksymalna (względem długości) spójna sekwencja sufiksów w $I$ które mają te same początkowe $h$ znaków to \textbf{grupa}
+    \item Grupa składająca się z co najmniej 2 sufiksów jest \textbf{grupą nieposortowaną}
+    \item Grupa składająca się z jednego sufiksu jest \textbf{grupą posortowaną}
+    \item Maksymalna spójna sekwencja posortowanych grup jest \textbf{wspólną grupą posortowaną}
+\end{itemize}
+\end{definition}
+Grupy będziemy numerować żeby móc przeprowadzać obliczenia na poszczególnych grupach z osobna. Dla grupy $I[f \ldots g]$ numer tej grupy to $g$. Ponadto tablica $V[i] = g$ zawiera informację, że sufiks $S_{i+1}$ znajduje się w grupie o numerze $g$. Dodatkowo będziemy korzystać z tablicy $L$ która trzyma długości nieposortowanych grup oraz długości wspólnych grup posortowanych. Dla rozróżnienia, długości tych późniejszych będą trzymane jako liczba ujemna. Tablicę $L$ użyjemy w podstawowej wersji algorytmu. Udoskonalenia omówione pod koniec tego omówienia pozwolą nam obyć się bez osobnej tablicy na długości grup.
+\begin{observation}
+Gdy I jest w \textit{h-order}, każdy sufiks w \textbf{wspólnej grupie posortowanej} jest unikalnie rozróżnialny od wszystkich innych sufiksów przy pomocy pierwszych $h$ symboli.
+\end{observation}
+Mając na uwadze powyższy niezmiennik, będziemy `przeskakiwać` nad grupami posortowanymi.
+
+Podstawowy algorytm będzie się prezentował następująco:
+\begin{enumerate}
+    \item Umieść sufiksy $1 \ldots n+1$ w tablicy $I$. Posortuj $I$ używając $x_i$ jako klucza dla sufiksu $i$ (pierwsza litera). Ustaw $h=1$
+    \item Dla każdego sufiksu $S_{i+1}$ ustaw numer grupy do której należy w $V[i]$
```

Comment:
For completeness add $L[i] = 1$, as you mention it later.

Reply:
I'm not sure I understand. I don't think setting $L[i] = 1$ belongs to this point. I mentioned initializing array L in next step of the algorithm.

Reply:
OK, that's fine